---
title: "RTNmotifs: An R/Bioconductor package for co-regulation analysis and inference of 'dual regulons'."
author: "Vinicius S. Chagas, Clarice S. Groeneveld, ..., Gordon Robertson, Mauro A. A. Castro"
date: "`r Sys.Date()`"
bibliography: bibliography.bib
output: BiocStyle::html_document
abstract: RTNmotifs is a spinoff tool derived from the RTN package, which infers regulons by mapping significant interactions between a list of regulators and candidate targets using Mutual Information (MI) (for details, please refer to the [RTN documentation](https://www.bioconductor.org/packages/devel/bioc/html/RTN.html)). The RTNmotifs searches for possible co-regulatory loops between regulon pairs. The tool takes two regulons and compares shared targets. All possible pairs are tested in order to infer "dual regulons", a new concept that tests whether regulons agree on the predicted downstream effects.
vignette: >
  %\VignetteIndexEntry{"RTNmotifs: An R/Bioconductor package for co-regulation analysis and inference of 'dual regulons'."}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\SweaveUTF8
---

# Overview

**RTNmotifs** searches for shared targets between regulators using different regulatory networks generated by the **RTN package** [@Castro2016]. Each regulator has a cloud of targets (i.e. regulon), and each target can be regulated in a positive or negative direction. Any shared target between a pair of regulators is regarded as a ‘motif’ and can be regulated in the same or in opposite directions (i.e. regulators can either agree or not on the predicted downstream effects). Because As a regulon comprises a considerable number of targets, many motifs can be predicted between two regulators. The ‘dual regulons’ represent regulon pairs whose common targets are potentially affected by both regulators. The number of shared targets is also assessed in order to test whether this number is greater than what would be expected by chance. The inference of ‘dual regulons’ is based on three complementary statistics: (1) the association between any two regulators assessed by MI, (2) the association between a given regulator and its potential targets, also assessed by MI (i.e. the regulon), and (3) the association between the set of motifs of any two regulons using correlation analysis. The schematics in **Figure 1** shows examples of ‘dual regulons’ with two motifs each. In (a) the two regulators agree on the downstream effects (i.e. same directions), while in (b) they disagree (i.e. opposite directions).

![title](figures/motifs.png)
<legend><b>Figure 1</b> - The 'duals regulons' possibilities and the respective downstream effects. The figure shows two schematics 'dual regulons' and the respective subnetworks showing the types of directions (downstream effect) in the motifs. In (<b>a</b>)  The ‘dual regulon’ agree in the effect (orange) and the filled dot represents that motif is repressed by both regulators while the unfilled dot shows that both regulators activate the motif. The opposite occurs in (<b>b</b>), in this type of ‘dual regulon’ the regulators disagree in the motifs downstream effect (green), the filled dot represents that the motif are repressed by the regulator 1 (R1) and activated by the regulator 2 (R2), whereas the unfilled dot shows the opposite. </legend>


# Quick Start
The **RTNmotifs** workflow starts with a preprocessing step that generates an **MBR-class** object from an expression matrix and two lists of regulators. The expression matrix is typically obtained from multiples samples (e.g. transcriptomes from a cancer cohort), while the lists of regulators represent some prior biological information indicating which genes in the expression matrix should be regarded as regulators. The input data can also deal with two classes of regulotors, as for example genes and microRNAs. In this case, the expression matrix should comprise mRNA and miRNA expression values. Alternativelly, the **MBR-class** object can be obtained by combining two **TNI-class** objects pre-computed in the **RTN** package.

## Load datasets
This example provides the data required to generate an **MBR-class** object. The dataset **dt4rtn** is available from the **RTN** package and consists of an R list with 6 objects, 3 of which will be used in the subsequent analysis: (1) **gexp**, a named gene expression matrix with 250 samples (genes in rows, samples in cols), (2) **gexpIDs**, a data.frame with Probe-to-ENTREZ annotation, and (3) **tfs**, a named vector listing several transcription factors. These datasets should be regarded as toy examples for demonstration purposes only, despite being extracted, pre-processed and size-reduced from @Fletcher2013 and @Curtis2012.

```{r include=FALSE}
library(RTNmotifs)
data("dt4rtn", package = "RTN")
gexp <- dt4rtn$gexp
annot <- dt4rtn$gexpIDs
tfs1 <- dt4rtn$tfs[c("IRF8","IRF1","PRDM1","AFF3","E2F3")]
tfs2 <- dt4rtn$tfs[c("HCLS1","STAT4","STAT1","LMO4","ZNF552")]
```

```{r eval=FALSE}
##---load package and prepare a dataset for demonstration
library(RTNmotifs)
data("dt4rtn", package = "RTN")
gexp <- dt4rtn$gexp
annot <- dt4rtn$gexpIDs
tfs1 <- dt4rtn$tfs[c("IRF8","IRF1","PRDM1","AFF3","E2F3")]
tfs2 <- dt4rtn$tfs[c("HCLS1","STAT4","STAT1","LMO4","ZNF552")]
```

## Preprocessing
The **gexp** data matrix and the corresponding annotation are evaluated by the `mbr.preprocess` function in order to check the consistency of the input data. After this step it is generated a pre-processed **MBR-class** object whose status is updated to 'Preprocess [x]'.

```{r eval=TRUE}
##---generate a pre-processed BR-class object
rmbr <- mbr.preprocess(gexp=gexp, regulatoryElements1=tfs1, regulatoryElements2=tfs2, gexpIDs=annot, verbose = FALSE)
rmbr
```

## Run the permutation analysis
The **mbr.permutation** method inherits the same algorithm implemented in the **RTN** package. This function takes the pre-processed **MBR-class** object and returns two regulatory networks inferred by mutual information analysis (with multiple hypothesis testing corrections). The results are included in 'TNI' slots, which will be used in the subsequent steps of the pipeline.

```{r eval=TRUE}
##---compute two regulatory networks
##---(set nPermutations>=1000)
rmbr <- mbr.permutation(rmbr, nPermutations=100, verbose = FALSE)
rmbr
```

## Run the bootstrap analysis
In additional to the permutation analysis, the stability of the regulatory networks can be assessed by bootstrapping using the `mbr.bootstrap` function, which also inherits the same algorithm from the **RTN** package. Each "TNI" slot of the **MBR-class** object is updated with a consensus bootstrap network.

```{r eval=TRUE}
##---check the stability of the two regulatory networks
##---(set nBootstrap>=100)
rmbr <- mbr.bootstrap(rmbr, nBootstrap=10, verbose=FALSE)
rmbr
```

## Run DPI filter
In a regulatory network each target can be linked to multiple regulators as a result of both direct and indirect interactions. The Data Processing Inequality (DPI) algorithm [@Meyer2008] is used to remove the weakest interaction of any triangle formed by two regulators and a common target, thus preserving the dominant regulator-target interaction. This step inherits the same algorithm implemented in the **RTN** package. It is an optional step in the context of the current pipeline, as it is applied to the same regulatory network.

```{r eval=TRUE}
##---apply DPI algorithm
rmbr <- mbr.dpi.filter(rmbr, eps=0.05, verbose=FALSE)
rmbr
```

## Run the association analysis between regulons
The `mbr.association` method takes the two transcriptional networks and cross check the regulons to search for motifs between all possible 'dual regulons'. The method computes the downstream motifs effects to check the agree between the possible 'dual regulons' and also the mutual information between the two regulators is evaluated. The correlation agree is computed for each 'dual regulon' and organized in quantiles, the qunatiles can be selected using the 'prob' parameter. A hypergeometric test evaluates if the common targets of a 'dual regulon' are potentially affect by both regulators in a number greater than what would be expected by chance. Each 'dual regulon' has computed the mutual information between the regulators and the regulons Jaccard coefficient of agreement and disagreement. The number of possible 'dual regulons' computed and the top and bottom correlation values can be accessed in the 'summary' slot using the `mbr.get` function. The `mbr.get` function can also accesses the table with the hypergeometric analysis values.

```{r eval=TRUE}
##---run the main RTNmotifs methods
rmbr <- mbr.association(rmbr, prob=0.75, verbose=FALSE)
rmbr
```

```{r eval=TRUE}
mbr.get(rmbr, what="summary")
```

## Organize the possible 'duals'
The `mbr.duals` method organizes the 'duals regulons' inferred by the `mbr.association` step using the correlation value. This method can add a complementary evidence when a 'supplementary.table' is provided. The 'supplementary.table' is a 'data.frame' listed with unique possible relationships between the regulatory elements used to compute the 'dual regulons' and the evidence of these relationships. This data needs to have three columns: (1) A column with the identifiers of the 'regulatoryElements1' input in the `mbr.preprocess`, (2) A column with the identifiers of the 'regulatoryElements2' input in the `mbr.preprocess` and (3) A column with the evidence value of these relationship. The name of the column evidence needs to be indentified in the input parameter 'evidenceColname'. This evidence information will be added in the 'motifsInformation' table.

```{r eval=TRUE}
##---make a toy example of additional evidences
motifsInformation <- mbr.get(rmbr, what="motifsInformation")
n <- nrow(motifsInformation)
idx <- sample(1:n, size=round(n/2))
supplementaryTable <- motifsInformation[idx,c("Regulon1","Regulon2")]
supplementaryTable$ToyEvidence <- rnorm(nrow(supplementaryTable))

##---run mbr.duals
rmbr <- mbr.duals(rmbr, supplementary.table = supplementaryTable, 
                  evidenceColname = "ToyEvidence", verbose = FALSE)
```

## Visualize the shared cloud of targets
The package provides a `mbr.dualsPlot` function to visualize the downstream effect that a 'dual regulon' have in the motifs. **Figure 2** shows two examples of 'dual regulon' plot, in **a** both regulators agree in the effect and in **b** the opposite can be visualized, the colors pattern is the same that was previously explained.  

```{r eval=FALSE}
dual <- mbr.get(rmbr, what="dualsRegulons")[1]
mbr.dualsPlot(rmbr, names.motifs = c(dual), alpha = 0.90)
```

![title](figures/duals.png)
<legend><b>Figure 2.</b> Two example of plots provided by the 'mbr.dual.plots' function. In (<b>a</b>) the plot profiling is equal that was expected in a 'dual regulon' that agree in the downstream effect in the motifs. In an other hand, (<b>b</b>) shows the pattern that was expected in a 'dual regulon' that disagree in the downstream effect in the motifs. The colours are in according to the pattern explained in <b>Figure 1</b>.</legend>

## Session Info
```{r}
sessionInfo()
```

## References
