---
title: "RTNmotifs: An R/Bioconductor package for co-regulation analysis and inference of 'dual regulons'."
author: "Vinicius S. Chagas, Clarice S. Groeneveld, Kerstin B Meyer, Gordon Robertson, Mauro A. A. Castro"
date: "`r Sys.Date()`"
bibliography: bibliography.bib
output: BiocStyle::html_document
abstract: Multiple regulators can co-operate or compete to control many biological processes. RTNmotifs is a tool derived from the [RTN](https://www.bioconductor.org/packages/devel/bioc/html/RTN.html) (Reconstruction of Transcriptional Networks) package and implemented to identify network motifs that involve pairs of regulators. Members of a pair can be of similar types or different types (e.g. transcription factor (TF)-TF, miRNA-miRNA, TF-miRNA). RTN infers interactions between a list of regulators and candidate target gene sets (i.e. regulons) using Mutual Information (MI). RTNmotifs searches for possible co-regulatory relationships between regulon pairs by assessing all possible network motifs (comprising two regulators and a common target) in an RTN regulatory network. It enumerates all possible regulon pairs in order to infer "dual regulons", a new concept that tests whether regulons co-regulate the predicted downstream targets. RTNmotifs can be applied to many different cellular regulatory processes. 
vignette: >
  %\VignetteIndexEntry{"RTNmotifs: An R/Bioconductor package for co-regulation analysis and inference of 'dual regulons'."}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\SweaveUTF8
---

# Overview
Recurrent patterns in biological networks have been termed network motifs [@Bracken2016], which may reflect critical roles in multiple biological processes; for example, regulatory loops between Transcription Factors and microRNAs [@Zhang2015]. **RTNmotifs** searches for targets shared between pairs of regulators, using regulatory networks generated by the RTN package (for details, please refer to the [RTN documentation](https://www.bioconductor.org/packages/devel/bioc/html/RTN.html)) [@Castro2016]. Each regulator has an associated set of gene targets (i.e. a regulon), and each target can be regulated in a positive or negative direction by a give regulator (i.e. pairs of regulators can either agree or disagree on the predicted downstream effects for a shared target gene). Because a regulon comprises a considerable number of targets, many network motifs can be predicted for a pair of regulators. ‘Dual regulons’ represent regulon pairs whose common targets are likely to be affected by both regulators. The number of shared targets is also assessed in order to test whether this number is greater than that expected by chance. The inference of 'dual regulons' requires three complementary statistics: (1) Targets are assigned to regulons based on MI between the regulator and the target. Significance of the MI statistics is assessed by permutation and bootstrap analysis. The association between pairs of regulators is also assessed in this step. (2) Shared targets between any two regulons are identified and the similarity in regulation is assessed by correlation analysis. Single network motifs are identified in this step. (3) A test is carried out to determine if the correlation between the set of network motifs of any two regulons is higher than would be expected by chance. The schematics in **Figure 1** show examples of ‘dual regulons’ with two network motifs each. In (a) the two regulators agree on the downstream effects (i.e. same directions), while in (b) they disagree (i.e. opposite directions). Our method can be applied to any regulatory relationship. For gene expression data sets typical regulators might include transcription factors, miRNAs, eRNAs and lncRNAs.

![title](figures/motifs.png)
<legend><b>Figure 1</b> -  Examples of regulators and predicted associations. This figure illustrates two ‘dual regulons’ and the respective downstream effects. (<b>a</b>)  Regulators R1 and R2 agree in both activation (open circle) and repression (filled circle). (<b>b</b>) Regulators R1 and R2 disagree. </legend>


# Quick Start
The **RTNmotifs** workflow starts with a preprocessing step that generates an **MBR-class** (Motifs Between Regulons) object from an expression matrix and two lists of regulators. The expression matrix is typically obtained from multiple samples (e.g. transcriptomes from a cancer cohort), while the lists of regulators represent some prior biological information indicating which genes in the expression matrix should be regarded as regulators. The input data can also deal with two classes of regulators; for example, genes and microRNAs. In this case, the expression matrix should comprise mRNA and miRNA expression values. Alternatively, the **MBR-class** object can be obtained by combining two **TNI-class** objects pre-computed in the **RTN** package.

## Load datasets
This example provides the data required to generate an **MBR-class** object. The dataset **dt4rtn** is available from the **RTN** package and consists of an R list with 6 objects, 3 of which will be used in the subsequent analysis: (1) **gexp**, a named gene expression matrix with 250 samples (genes in rows, samples in cols), (2) **gexpIDs**, a data.frame with Probe-to-ENTREZ annotation, and (3) **tfs**, a named vector listing transcription factors (in this case, 149 TFs). While these datasets were extracted, pre-processed and size-reduced from @Fletcher2013 and @Curtis2012, they should be regarded as examples for demonstration purposes only

```{r include=FALSE}
library(RTNmotifs)
data("dt4rtn", package = "RTN")
gexp <- dt4rtn$gexp
annot <- dt4rtn$gexpIDs
tfs1 <- dt4rtn$tfs[c("IRF8","IRF1","PRDM1","AFF3","E2F3")]
tfs2 <- dt4rtn$tfs[c("HCLS1","STAT4","STAT1","LMO4","ZNF552")]
```

```{r eval=FALSE}
##---load package and prepare a dataset for demonstration
library(RTNmotifs)
data("dt4rtn", package = "RTN")
gexp <- dt4rtn$gexp
annot <- dt4rtn$gexpIDs
tfs1 <- dt4rtn$tfs[c("IRF8","IRF1","PRDM1","AFF3","E2F3")]
tfs2 <- dt4rtn$tfs[c("HCLS1","STAT4","STAT1","LMO4","ZNF552")]
```

## Preprocessing
The **gexp** data matrix and the corresponding annotation are evaluated by the `mbr.preprocess` function in order to check the consistency of the input data. After this step it is generated a pre-processed **MBR-class** object whose status is updated to 'Preprocess [x]'.

```{r eval=TRUE}
##---generate a pre-processed BR-class object
rmbr <- mbr.preprocess(gexp=gexp, regulatoryElements1=tfs1, regulatoryElements2=tfs2, gexpIDs=annot, verbose = FALSE)
rmbr
```

## Run the permutation analysis
The `mbr.permutation` method inherits the same algorithm implemented in the **RTN** package. This function takes the pre-processed **MBR-class** object and returns two regulatory networks that are inferred by mutual information analysis (with multiple hypothesis testing corrections). The results are included in the 'TNI' slots, which will be used in the subsequent steps of the pipeline.

```{r eval=TRUE}
##---compute two regulatory networks
##---(set nPermutations>=1000)
rmbr <- mbr.permutation(rmbr, nPermutations=100, verbose = FALSE)
rmbr
```

## Run the bootstrap analysis
In additional to the permutation analysis, the stability of the regulatory networks is assessed by bootstrapping using the `mbr.bootstrap` function, which also inherits the same algorithm from the **RTN** package. Each 'TNI' slot of the **MBR-class** object is updated with a consensus bootstrap network.

```{r eval=TRUE}
##---check the stability of the two regulatory networks
##---(set nBootstrap>=100)
rmbr <- mbr.bootstrap(rmbr, nBootstrap=10, verbose=FALSE)
rmbr
```

## Run DPI filter
In a given regulatory network each target can be linked to multiple regulators as a result of both direct and indirect interactions. The Data Processing Inequality (DPI) algorithm [@Meyer2008] is used to remove the weakest interaction between two regulators and a common target. This step inherits the algorithm that is implemented in the **RTN** package, and it is optional for the analyses described in this workflow (the results of this step can be used to assess regulon activity in the **RTN** package).

```{r eval=TRUE}
##---apply DPI algorithm
rmbr <- mbr.dpi.filter(rmbr, eps=0.05, verbose=FALSE)
rmbr
```

## Run the association analysis between regulons
The `mbr.association` method takes the two transcriptional networks computed in the previous steps and enumerates all motifs between all regulons. The method retrieves the mutual information between regulators and assesses the agreement between the predicted downstream effects using correlation analysis. A hypergeometric test is used to evaluate whether the common targets are potentially affected by both regulators in a number greater than what would be expected by chance. A summary of the results can be accessed from the 'summary' slot using the `mbr_get` function.
```{r eval=TRUE}
##---run the main RTNmotifs methods
rmbr <- mbr.association(rmbr, prob=0.75, verbose=FALSE)
rmbr
```

```{r eval=TRUE}
mbr_get(rmbr, what="summary")
```

## Rank 'dual regulons'
The `mbr.duals` method ranks all candidates using the correlation values computed in the `mbr.association` step, and returns the list of 'dual regulons'. Also, when a supplementary.table is available this method can add complementary evidence regarding the association between regulators. The supplementary.table is a 'data.frame' listing unique relationships between any two regulators used to compute the 'dual regulons' (please refer to the documentation for details on the input data format).

```{r eval=TRUE}
##---make a toy example of additional evidences
motifsInformation <- mbr_get(rmbr, what="motifsInformation")
n <- nrow(motifsInformation)
supplementaryTable <- motifsInformation[1:n,c("Regulon1","Regulon2")]
supplementaryTable$ToyEvidence <- rnorm(n)

##---run mbr.duals
rmbr <- mbr.duals(rmbr, supplementary.table = supplementaryTable, 
                  evidenceColname = "ToyEvidence", verbose = FALSE)
```

## Visualize the shared cloud of targets
The package provides the `mbr.plot.duals` function to visualize 'dual regulon' and the respective motifs, which represent the shared cloud of targets. **Figure 2** shows two examples of 'dual regulons' for the schematics illustrated in **Figure 1**. In **(a)** the two regulators agree on the downstream effect (i.e. same directions), while in **(b)** they disagree (i.e. opposite directions).

```{r eval=FALSE}
dual <- mbr_get(rmbr, what="dualRegulons")[1]
mbr.plot.duals(rmbr, names.motifs = c(dual), alpha = 0.90)
```

![title](figures/duals.png)
<legend><b>Figure 2.</b>  The <b>'mbr.plot.duals'</b> function shows the shared cloud of targets. In (<b>a</b>) the regulators agree on the downstream effects, while in (<b>b</b>) they disagree. The colour pattern follows the schematics in <b>Figure 1</b>.</legend>

## Session Info
```{r}
sessionInfo()
```

## References
